fix: [BUG]: Backend coverage report is not helpful for debugging

Ykumar1415 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #19797.
2. This PR does the following: Remove the coverage details of the files having 100% coverage.
3. (For bug-fixing PRs only) The original bug occurred because: [Explain what
   the cause of the bug was, and which PR introduced it]

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [x] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)

## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->
- [ ] A testing doc has been written: ... (ADD LINK) ...
- [ ] _(To be confirmed by the server admin)_ All jobs in this PR have been verified on a live server, and the PR is safe to merge.


## Proof that changes are correct
![Screenshot from 2024-03-21 02-03-18](https://github.com/oppia/oppia/assets/97088265/f6d2c099-d367-4fa8-8768-967a2d55cdde)
![Screenshot from 2024-03-21 02-03-38](https://github.com/oppia/oppia/assets/97088265/0abbe870-9745-476f-8f02-97ad640115a4)

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Ykumar1415 committed:
backend coverage result

Ykumar1415 committed:
some changes

Ykumar1415 committed:
Update lint checks.

Ykumar1415 commented:
@U8NWXD can you review this pr?

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -603,7 +603,43 @@ def main(args: Optional[List[str]] = None) -> None:
+                'WARNING: Backend test coverage is below 100%.'
+                ' The rightmost "Missing" column above '
+                'shows which lines are uncovered.'
+                )
```
Deindent by 4, similarly in 2 cases below.

Ykumar1415 replied:
@seanlip i have deindent these lines and some other lines( print statement below) changes in new updated commit.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -659,6 +695,25 @@ def check_coverage(
         check=False)
+
+    not_covered_lines = []
+    is_first_line = True
```
Try not to start booleans with a verb. Maybe something_is_first_line (specify what "something" is).

seanlip replied:
Also do you even need this (couldn't you just check whether not_covered_lines is empty)?

Ykumar1415 replied:
@seanlip We need this variable in the case where all the files have 100% coverage. In such a scenario, there is no need to add the heading.

Ykumar1415 replied:
@seanlip i have removed the boolean variable in new updated code. Thanks.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -659,6 +695,25 @@ def check_coverage(
+    output_lines = process.stdout.split('\n')
+
+    for index, line in enumerate(output_lines):
+        if line and (' 100%' not in line and '-----' not in line
```
What are you trying to search for here? Maybe add a comment to explain, or better, add a new named variable after `if line` that stores the result of `all(fragment not in line for fragment in ['100%', '-----', 'Name'])` that describes what that computation represents.

Ykumar1415 replied:
@seanlip In the coverage report, the first line consists of the table headers, such as "Name," "Missing," etc. The second line is a row separator containing dashes (----).

For files with coverage less than 100%, the details are stored in a variable called "not_covered_lines." This variable contains the heading followed by the coverage report of each incomplete file, separated by a dash line (---).

that is the complete thing that I am doing here. I will add good naming of the variable here. Thanks.

Ykumar1415 replied:
@seanlip can you check the code below  : 

![WhatsApp Image 2024-03-07 at 01 56 48_007e6c9f](https://github.com/oppia/oppia/assets/97088265/944c299e-576a-4eaa-93dd-740b3312e189)

seanlip replied:
Thanks, this is a bit better. Some thoughts:
- Why is your 'if' check so complicated ... would `` 100%` not in line` be enough? As mentioned, try to separate out that clause into its own variable and name it properly and put a comment on it to explain why all those checks are being made.
- `all_files_covered` is only set to False when something is added to `incomplete_coverage_lines`. So why can't you just use `if incomplete_coverage_lines` to check whether all files are covered or not?
- Perhaps some comment that explains the format of process.stdout might be useful, so that the reader has a basis to analyze the code that parses that info.

Ykumar1415 replied:
![image](https://github.com/oppia/oppia/assets/97088265/96030f5c-45a2-4168-b4c0-1c37e8edefd7)
@seanlip  as suggested i have included the comments to understand better what this code is doing. 
 I removed the all_files_covered flag from the code. 

below is the format of the process.stdout which will help to understand the working of the code: 
![image](https://github.com/oppia/oppia/assets/97088265/4254c813-de39-49ad-a6de-c2e8d4a884f2)

seanlip replied:
Put a long comment in the file that explains the current format so that devs modifying this part of the code can understand it without needing to refer to other sources. It is fine if your comment is multiple paragraphs, just make sure it is clear.

Ykumar1415 replied:
![image](https://github.com/oppia/oppia/assets/97088265/0cd95d8c-514f-4fd0-8235-b2cf0feca6c2)
@seanlip Please review the attached screenshot for changes made to the code.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -659,6 +695,25 @@ def check_coverage(
+    for index, line in enumerate(output_lines):
+        if line and (' 100%' not in line and '-----' not in line
+         and 'Name' not in line):
+            if is_first_line and index > 0:
```
Not sure what `index` refers to here. Maybe give it a more specific name?

Ykumar1415 replied:
@seanlip i included correct name as suggested by you in the updated code. Thanks.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -659,6 +695,25 @@ def check_coverage(
         cmd, capture_output=True, encoding='utf-8', env=env,
         check=False)
+
+    not_covered_lines = []
```
I don't understand what this is storing, perhaps add a comment and improve the naming.

Ykumar1415 replied:
@seanlip i tried to explain this in updated code.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -659,6 +695,25 @@ def check_coverage(
+                not_covered_lines.append(output_lines[1])
+                is_first_line = False
+            not_covered_lines.append(line)
+            i = index
```
Don't know what i is, please improve the naming to explain what concept you are representing here.

Ykumar1415 replied:
@seanlip i removed i in updated code.

seanlip commented on scripts/run_backend_tests_test.py:
```diff
@@ -916,8 +930,10 @@ def mock_subprocess_run(cmd: List[str], **_: str) -> MockProcessOutput:
+        expected_output = coverage_report_output.strip()
+        returned_res = returned_output.strip()
+        self.assertEqual(expected_output, returned_res)
+        self.assertEqual(coverage, 62.0)
```
I think there should be at least one unit test that verifies the log that is displayed for the end user (the developer in this case). Maybe two tests for the cases "there is a coverage gap to report" and "there is no coverage gap to report".

Ykumar1415 replied:
@seanlip does coverage gap mean the condition when all the files will have complete coverage?

seanlip replied:
A "coverage gap" means that some file has a lack of coverage.

Ykumar1415 replied:
@seanlip I have one last question. The test case I added currently focuses on scenarios where there is a coverage gap (files lack coverage). I believe we can enhance it by combining it with another test case, where some files have complete coverage and others have a lack of coverage. This way, we can test both scenarios within a single test case. Does this approach sound correct?

seanlip replied:
I suggest you have one test where there's a coverage gap, one test where there's no coverage gap, and one test that mixes the two. 

It is good to have focused test cases since that will make it easier to identify when an issue occurs. With a single test case that covers multiple scenarios, diagnosing issues becomes harder.

Ykumar1415 replied:
@seanlip I have included tests for all three conditions in first there are only uncovered files, second is then there are only covered lines and third is when both covered and uncovered files will be present in coverage report. Thanks.

seanlip reviewed (changes requested):
Thanks @Ykumar1415! Left some notes.

U8NWXD commented on scripts/run_backend_tests.py:
```diff
@@ -603,7 +603,43 @@ def main(args: Optional[List[str]] = None) -> None:
         report_stdout, coverage = check_coverage(True)
+        print('')
+        print(
+            '+-------------------------------------------------------------'
+            '--------------------------------------------------------------'
+            '-------+'
```
Please limit your output lines to 80 characters in length so that they stay readable on narrow terminal windows (e.g. when developers have a terminal side-by-side with their code editor)

Ykumar1415 replied:
ok @U8NWXD i will make it only 80 character long line.

U8NWXD commented on scripts/run_backend_tests.py:
```diff
@@ -603,7 +603,43 @@ def main(args: Optional[List[str]] = None) -> None:
     if parsed_args.generate_coverage_report:
```
How do the coverage issues raised here compare to those reflected in the `incomplete_coverage` variable above? When `incomplete_coverage != 0`, we raise an exception and won't ever reach the code you added.

Ykumar1415 replied:
@U8NWXD In this specific line, we retrieve the incomplete_coverage value from the following line of code:
`check_test_results(tasks, task_to_taskspec, parsed_args.generate_coverage_report)`

Within the check_test_results function, we increase the incomplete_coverage under two conditions:
`spec.test_target not in coverage_exclusions and float(coverage) != 100.0`

when we see the 4th backend test shard the **spec.test_target** value is "**core.jobs.batch_jobs.subtopic_migration_jobs_test,**" which is not listed in the **coverage_exclusions**. However, the coverage for this particular file is **100%**. As a result, the **incomplete_coverage** value is **0**.

The same situation is occurring in the other four shards as well, leading to an **incomplete_coverage** value of **0** for all four shards.

U8NWXD replied:
@Ykumar1415 for a given shard, `spec.test_target` takes on multiple values, namely the test module for each test in the shard. Here's the loop where this iteration happens:

https://github.com/oppia/oppia/blob/b210922fff2e55c705342a204955f86f960bc286/scripts/run_backend_tests.py#L379-L381

So in the 4th shard, I don't think `spec.test_target` only takes on the value `core.jobs.batch_jobs.subtopic_migration_jobs_test`.

That said, I think your answer clarifies things for me. There are two ways we enforce coverage:
1. For a file not in the `scripts/backend_tests_incomplete_coverage.txt` exclusion list, we require that its associated test file provide 100% coverage by itself. This calculation ignores any coverage that might be provided by other tests.
2. For all files, even those in the exclusion list, we require that all the tests combined give 100% coverage. Some of this coverage can come from tests other than the associated test file.

Your updates only apply when there are no coverage errors of the first type, but there are errors of the second type. Does this match your understanding? If so, I think that could be more clearly communicated in your outputs.

seanlip replied:
Just a note -- let's ignore (1) for the purposes of making the coverage log clearer. We should focus on (2). I've had a chat with @U8NWXD about removing the functionality for (1) since we are not planning to enforce it.

Ykumar1415 replied:
Ok @seanlip

Ykumar1415 committed:
added tests for different cases

Ykumar1415 committed:
added tests for different cases

Ykumar1415 committed:
backend test included

U8NWXD commented on scripts/run_backend_tests.py:
```diff
@@ -603,7 +603,39 @@ def main(args: Optional[List[str]] = None) -> None:
+                'lines in each file.'
+                )
+
+            print(
+                'For more information, please see the '
+                '[backend tests wiki page]'
+                '(https://github.com/oppia/oppia/wiki/'
+                'Backend-tests#coverage-reports ).'
+                )
```
There are still lines longer than 80 characters, for example here. When reviewers flag problems, please be sure to correct all instances of the problem, not just the flagged one. We often only flag one occurrence of an issue to avoid spamming PRs with review comments

Ykumar1415 replied:
@U8NWXD i will remember it in future. I have one doubt that should i make each 80 character to be printed in one line ?.
And regarding #19886 @seanlip told to leave the condition -1 (in comment)

U8NWXD replied:
There are two requirements:
1. No line in the code should exceed 80 characters in length
2. No line in the output should exceed 80 characters in length

So to answer your question directly, no, you should not print an 80-character line using one line of code if that line of code would exceed 80 characters. You should still break lines of code across multiple lines as you've already done. Alternatively, you could make your output lines even shorter so that the code printing them fits on one line.

Ykumar1415 replied:
@U8NWXD Thanks. now I understand the issues with my code. 
In the previous version, there was no restriction on the length of warning statements in the output. Now, I've corrected the print statement to ensure that no lines exceed 80 characters. Additionally, I've reduced the level of indentation as suggested by Sean.

U8NWXD reviewed (changes requested):
Almost there! Just one more comment from me, plus https://github.com/oppia/oppia/pull/19886#discussion_r1517097599 from before, which is still open

Ykumar1415 commented:
@U8NWXD i will remember it in future. I have one doubt that should i make each 80 character to be printed in one line ?.
And regarding https://github.com/oppia/oppia/pull/19886 @seanlip told to leave the condition -1 (in comment)

seanlip commented:
Also relatedly ... I'm assigning this back since my comments haven't been replied to yet (at least, from my perspective).

Ykumar1415 commented:
> Also relatedly ... I'm assigning this back since my comments haven't been replied to yet (at least, from my perspective).

@seanlip   Now that I've implemented the process you suggested, my comments are visible now (confirmed in incognito mode). Thank you.

Ykumar1415 commented:
@seanlip @U8NWXD  I have one doubt that should i make each 80 character to be printed in one line ? because in codebase no line exceeds 80 characters.

Ykumar1415 committed:
corrected output character related problem

Ykumar1415 committed:
Merge branch 'develop' into coverage-report

Ykumar1415 commented:
> Thanks @Ykumar1415! Left some notes.



> I'm not sure why I'm assigned, not all my comments have been replied to. Please respond to and address **all** comments before requesting a re-review, thanks.

@seanlip I believe I have addressed all the comments except for one change made in the latest commit, which I overlooked in my response. I  will ensure to verify all comments thoroughly in the future.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -603,7 +603,39 @@ def main(args: Optional[List[str]] = None) -> None:
         print(report_stdout)
+        if coverage != 100:
+            print(
+            'WARNING: Backend test coverage is below 100%.\n'
```
Indent this and the next 2 lines by 4, since you're creating a new "scope" with the open parens. Ditto for the next two print statements.

Ykumar1415 replied:
@seanlip i have made these change in new commit. But waiting for your reply for [this](https://github.com/oppia/oppia/pull/19886#discussion_r1527197909) comment . so that I can make commit having all the changes.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -603,7 +603,39 @@ def main(args: Optional[List[str]] = None) -> None:
+            print(
+            'WARNING: Backend test coverage is below 100%.\n'
+            'The rightmost "Missing" column above'
+            ' shows which lines uncovered.'
```
...lines are still uncovered.

Ykumar1415 replied:
@seanlip i have made this change in new commit. But waiting for your reply for [this](https://github.com/oppia/oppia/pull/19886#discussion_r1527197909) comment . so that I can make commit having all the changes.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -603,7 +603,39 @@ def main(args: Optional[List[str]] = None) -> None:
+
+            print(
+            'For more information, please see\n'
+            '[backend tests wiki page]\n'
```
please see our [...]

Ykumar1415 replied:
@seanlip i have made this change in new commit. But waiting for your reply for [this](https://github.com/oppia/oppia/pull/19886#discussion_r1527197909) comment . so that I can make commit having all the changes.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -603,7 +603,39 @@ def main(args: Optional[List[str]] = None) -> None:
+            'For more information, please see\n'
+            '[backend tests wiki page]\n'
+            '(https://github.com/oppia/oppia/wiki/'
+            'Backend-tests#coverage-reports ).'
```
Drop the space before the closing paren.

Ykumar1415 replied:
@seanlip i have made this change in new commit. But waiting for your reply for [this](https://github.com/oppia/oppia/pull/19886#discussion_r1527197909) comment . so that I can make commit having all the changes.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -659,6 +691,44 @@ def check_coverage(
+    # -------------------------------------------------
+    # TOTAL        30    0      6      0   95%.
+    incomplete_coverage_lines: List[str] = []
+    # Lines with incomplete coverage are appenede to incomplete_coverage_lines.
```
Spelling: appended

Ykumar1415 replied:
@seanlip i have made this change in new commit. But waiting for your reply for [this](https://github.com/oppia/oppia/pull/19886#discussion_r1527197909) comment . so that I can make commit having all the changes.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -659,6 +691,44 @@ def check_coverage(
+        # The line having incomplete coverage, not a dashed line(-----),
+        # and not the heading line is appended to incomplete_coverage_lines.
+
+        line_missing_full_coverage = line and (
```
line_is_missing_full_coverage

(this is more clearly a boolean, otherwise it sounds like it stores the line)

Ykumar1415 replied:
@seanlip i have made this change in new commit. But waiting for your reply for [this](https://github.com/oppia/oppia/pull/19886#discussion_r1527197909) comment . so that I can make commit having all the changes.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -659,6 +691,44 @@ def check_coverage(
+
+        # The line having incomplete coverage, not a dashed line(-----),
+        # and not the heading line is appended to incomplete_coverage_lines.
+
```
Drop this newline, the comment should be directly above the line it is for.

Ykumar1415 replied:
@seanlip i have made this change in new commit. But waiting for your reply for [this](https://github.com/oppia/oppia/pull/19886#discussion_r1527197909) comment . so that I can make commit having all the changes.

seanlip commented on scripts/run_backend_tests_test.py:
```diff
@@ -894,11 +894,45 @@ def mock_subprocess_run(cmd: List[str], **_: str) -> MockProcessOutput:
         self.assertEqual(coverage, 86)
 
-    def test_coverage_is_calculated_correctly_for_a_single_file(self) -> None:
+    def test_coverage_calculated_correctly_incomplete_coverage(self) -> None:
```
Looking at this, I have a question. Can't you just use the --skip-covered flag when generating the coverage report using coverage? https://stackoverflow.com/questions/20543981/how-to-omit-reporting-files-that-have-100-coverage-in-coverage-py

Ykumar1415 replied:
@seanlip the --skip-covered flag is valid for pytest Coverage.py 4.3.2 version. But in oppia, we are using unittest for writing backend tests so we can't use it.

seanlip replied:
Have you actually tried it? The official coverage docs also talk about this argument. https://coverage.readthedocs.io/en/7.4.0/cmd.html#coverage-summary-coverage-report

Ykumar1415 replied:
@seanlip Yes I tried it but the incorrect command passed error is showing. 
i passed the command with --skip-covered flag.
![image](https://github.com/oppia/oppia/assets/97088265/6b5780f8-316f-4bf9-a0bc-4a06500df0be)

seanlip replied:
That sounds to me like it might be more of an issue with your code, TBH. What does your call to subprocess.run() look like? What happens if you run the existing and revised commands directly from the terminal?

Ykumar1415 replied:
@seanlip this error is due to tests failure I am currently debugging this. Because when i am manually running directly on terminal it is working correctly.

Ykumar1415 replied:
yes, @seanlip, the issue stemmed from my oversight in updating run_backend_test.py without reflecting the changes in the test file. As a result, this error occurred. However, I've rectified the situation now. Consequently, we can proceed by omitting the code responsible for extracting lines with incomplete coverage.
Thanks.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -659,6 +691,44 @@ def check_coverage(
+            ' 100%' not in line and '-----' not in line and 'Name' not in line
+        )
+        if line_missing_full_coverage:
+            # If a line indicates incomplete coverage then Heading,
```
I don't really understand this. I think that, for conceptual clarity, `incomplete_coverage_lines` should only contain the actual data, i.e. lines with files with a percentage that's not 100.

All the other formatting can be added when you print the results out.

Ykumar1415 replied:
@seanlip Currently, the coverage report is generated within this function, and by performing the formatting at this stage, I avoid the need for additional code duplication and complexity elsewhere (because there one more looping will be needed ). So i added formatting also on the time of getting lines having coverage less than 100%.

Would it be beneficial to consider formatting the report output in terms of complexity?

seanlip replied:
I think the problem is that I can't understand what you are doing from reading the code.

Variables should clearly describe what you are storing in them, and comments should be added as needed to explain. When you say incomplete_coverage_lines I take that to mean exactly the lines that are not completely covered and nothing else. If you are including other stuff then you should call it something else. If you have difficulty coming up with a suitable name then that probably indicates you should use more easy-to-understand concepts instead, even if that means adding more logic.

In general: be clear about what your variables store, name them accordingly, and write the implementation to store exactly the right thing in each variable based on its name.

Ykumar1415 replied:
> I think the problem is that I can't understand what you are doing from reading the code.
> 
> Variables should clearly describe what you are storing in them, and comments should be added as needed to explain. When you say incomplete_coverage_lines I take that to mean exactly the lines that are not completely covered and nothing else. If you are including other stuff then you should call it something else. If you have difficulty coming up with a suitable name then that probably indicates you should use more easy-to-understand concepts instead, even if that means adding more logic.
> 
> In general: be clear about what your variables store, name them accordingly, and write the implementation to store exactly the right thing in each variable based on its name.

 @seanlip, I'm attempting to refactor the code to enhance its readability, making it comprehensible for everyone to understand its functionality.

Ykumar1415 replied:
@seanlip I've introduced a new function for improved code clarity, but I'm concerned it might not be easy to understand. -
![image](https://github.com/oppia/oppia/assets/97088265/969e74a0-c5b8-4c49-882b-ac6f823bb200)

seanlip commented on scripts/run_backend_tests_test.py:
```diff
@@ -916,8 +950,92 @@ def mock_subprocess_run(cmd: List[str], **_: str) -> MockProcessOutput:
 
-        self.assertEqual(returned_output, coverage_report_output)
-        self.assertEqual(coverage, 86)
+        expected_output = expected_coverage_report.strip()
```
Here and elsewhere, why do you need the strip()? Can you just update the string constants so that they can be used as-is?

Ykumar1415 replied:
@seanlip, the utilization of the strip() method within this context serves to eliminate leading and trailing whitespace characters from the strings. The discrepancy observed between the expected output and the output obtained from the check_coverage function is attributed to the presence of whitespace, thus necessitating the inclusion of .strip(). This ensures consistency and accuracy in the comparison process by disregarding extraneous whitespace.

seanlip replied:
Well, why not drop the extraneous whitespace from the expected string? You control that string completely, after all.

Ykumar1415 replied:
@seanlip now removed the .split() from test code.

seanlip reviewed (changes requested):
Thanks @Ykumar1415 -- took another pass. Most comments are small but PTAL at the top comment in the test file since looking into that might make all this a lot simpler.

U8NWXD reviewed (approved):
LGTM aside from seanlip's comments

Ykumar1415 committed:
suggested changes completed

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -648,7 +716,8 @@ def check_coverage(
          '--omit="%s*","third_party/*","/usr/share/*"'
-         % common.OPPIA_TOOLS_DIR, '--show-missing']
+         % common.OPPIA_TOOLS_DIR, '--show-missing',
+         '--skip-covered']
```
When you run coverage with this command, does the resulting output automatically skip 100% lines? (What does it look like, can you show an example?)

If it already skips "100% lines", then do you need all the complicated formatting still? Or does simply putting a message at the end suffice?

Ykumar1415 replied:
@seanlip i have included the screenshots of lines with formatting and without formatting. both don't have lines with 100%coverage. I think lines with formatting look less complicated.
**with formatting :** 
![image](https://github.com/oppia/oppia/assets/97088265/e946df89-f32f-41d9-b4bb-9ba53ddaf1ae)
![image](https://github.com/oppia/oppia/assets/97088265/f425d9fd-5472-4ec5-b932-99f2dbd96717)

**without formatting:** 
![image](https://github.com/oppia/oppia/assets/97088265/7eff184e-d490-4777-b7df-4ade0762835b)

seanlip replied:
Thanks for the screenshots. I agree the formatting looks a bit nicer, but typically we won't run into this case where so many lines are uncovered, it'll be just one or two files. 

Remember the original point of this issue -- show only the files that miss coverage, make clear what specific actions the dev should take next. The first part has been solved by the changed command. So just do what's needed to achieve the second part -- I don't think the complexity of the code needed for the pretty-printing is worth it.

Ykumar1415 replied:
ok @seanlip i will remove the formatting and then update the commit.

Ykumar1415 replied:
@seanlip i removed the code that was formatting the report and updated the test for the condition.

Ykumar1415 committed:
suggested changes completed

Ykumar1415 committed:
coverage-report-print problem resolved.

seanlip commented on scripts/run_backend_tests_test.py:
```diff
@@ -916,8 +925,10 @@ def mock_subprocess_run(cmd: List[str], **_: str) -> MockProcessOutput:
-        self.assertEqual(returned_output, coverage_report_output)
-        self.assertEqual(coverage, 86)
+        expected_output = coverage_report_output
+        returned_res = returned_output
```
Why rename this? Drop this and the line above and just use the original names.

Ykumar1415 replied:
ok @seanlip i removed these lines.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -603,7 +603,39 @@ def main(args: Optional[List[str]] = None) -> None:
+
+            print(
+                'For more information, please see our\n'
+                '[backend tests wiki page]\n'
```
Why do you need an \n at the end of this?

Ykumar1415 replied:
@seanlip I included ' \n' here, following @U8NWXD's recommendation to ensure that no lines in the code or output exceed 80 characters.

seanlip replied:
Then this doesn't seem to be reflected in your proof of changes? Please keep the proof of changes in your PR description up to date.

Ykumar1415 replied:
@seanlip i have updated the proof of changes. Thanks.

seanlip reviewed (changes requested):
Thanks @Ykumar1415, looks good. Just a couple notes.

Ykumar1415 committed:
minor change in  code.

Ykumar1415 committed:
some changes run_backend_tests_test file

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -603,7 +603,39 @@ def main(args: Optional[List[str]] = None) -> None:
+        print('')
         print(report_stdout)
+        if coverage != 100:
+            print(
```
@Ykumar1415 One more request, sorry. Could you please write the message like this:

```
print('WARNING: Backend test coverage is below 100%.')
print('')

print('The rightmost "Missing" column above shows lines that are still uncovered.')
print('Please add tests for scenarios that exercise those lines of code so that ')
print('there are no uncovered lines in each file.')
print('')

print('For more information, please see our backend tests wiki page:')
print('    https://github.com/oppia/oppia/wiki/Backend-tests#coverage-reports')
```

I think it would be easier to read -- take a look at the output produced and see if you agree. Thanks.

Ykumar1415 replied:
@seanlip i agree with you. These lines will be printed like : 
![image](https://github.com/oppia/oppia/assets/97088265/53ba6aa4-0d4b-41c3-9dda-baef66999ea8)

To adhere to our code linting rule, which limits the number of characters per line, we must structure the code accordingly to achieve the same output. code : 
![image](https://github.com/oppia/oppia/assets/97088265/ddfe059f-2bdf-43af-a871-de9802d29fdd)

But output will be same. i think it is easy to read at the output. Can i update Pr ?.

seanlip replied:
Yes that's fine (though your first screenshot doesn't show the full thing). Just note that in your print statements, the closing paren should match the indentation level of the line with the opening paren (i.e. deindent by 4 from what you have).

Ykumar1415 replied:
@seanlip i have updated the pr.

Ykumar1415 committed:
added statements for warning.

seanlip reviewed (approved):
Great. LGTM, thanks!

Ykumar1415 commented:
> Great. LGTM, thanks!

Thanks to @seanlip, working through this problem has been a valuable learning experience for me.

Ykumar1415 committed:
Merge branch 'develop' into coverage-report
